src/include/storage/copier/node_copier.h

```diff
@@ -1,136 +1,235 @@
-
-public:
-    NodeCopier(common::CopyDescription& copyDescription, std::string outputDirectory,
-        common::TaskScheduler& taskScheduler, catalog::Catalog& catalog, common::table_id_t tableID,
-        NodesStatisticsAndDeletedIDs* nodesStatisticsAndDeletedIDs)
-        : TableCopier{copyDescription, std::move(outputDirectory), taskScheduler, catalog, tableID,
-              nodesStatisticsAndDeletedIDs} {}
+    virtual std::unique_ptr<NodeCopier> clone() const {
+        return std::make_unique<NodeCopier>(sharedState, pkIndex, copyDesc, columns, pkColumnID);
+    }
 
 protected:
-    void initializeColumnsAndLists() override;
-
-    void populateColumnsAndLists(processor::ExecutionContext* executionContext) override;
-
-    void saveToFile() override;
+    virtual void executeInternal(std::unique_ptr<NodeCopyMorsel> morsel) {
+        throw std::runtime_error("Not implemented");
+    }
 
-    template<typename T>
+    static void copyArrayIntoColumnChunk(InMemColumnChunk* columnChunk, InMemNodeColumn* column,
+        arrow::Array& arrowArray, common::offset_t startNodeOffset,
+        common::CopyDescription& copyDescription, PageByteCursor& overflowCursor);
     static void populatePKIndex(InMemColumnChunk* chunk, InMemOverflowFile* overflowFile,
         common::NullMask* nullMask, HashIndexBuilder<T>* pkIndex, common::offset_t startOffset,
         uint64_t numValues);
+    static void appendPKIndex(InMemColumnChunk* chunk, InMemOverflowFile* overflowFile,
+        common::offset_t offset, HashIndexBuilder<T>* pkIndex) {
+        assert(false);
+    }
 
-    std::unordered_map<common::property_id_t, std::unique_ptr<InMemNodeColumn>> columns;
+protected:
+    std::shared_ptr<NodeCopySharedState> sharedState;
+    HashIndexBuilder<T>* pkIndex;
+    common::CopyDescription copyDesc;
+    std::vector<std::unique_ptr<InMemNodeColumn>>& columns;
+    common::column_id_t pkColumnID;
+    std::vector<PageByteCursor> overflowCursors;
+};
+
+template<>
+void NodeCopier<int64_t>::appendPKIndex(InMemColumnChunk* chunk, InMemOverflowFile* overflowFile,
+    common::offset_t offset, HashIndexBuilder<int64_t>* pkIndex);
+template<>
+void NodeCopier<common::ku_string_t>::appendPKIndex(InMemColumnChunk* chunk,
+    InMemOverflowFile* overflowFile, common::offset_t offset,
+    HashIndexBuilder<common::ku_string_t>* pkIndex);
+
+template<typename T>
+class CSVNodeCopier : public NodeCopier<T> {
+public:
+    CSVNodeCopier(std::shared_ptr<NodeCopySharedState> sharedState, HashIndexBuilder<T>* pkIndex,
+        const common::CopyDescription& copyDesc,
+        std::vector<std::unique_ptr<InMemNodeColumn>>& columns, common::column_id_t pkColumnID)
+        : NodeCopier<T>{std::move(sharedState), pkIndex, copyDesc, columns, pkColumnID} {}
+
+    inline std::unique_ptr<NodeCopier<T>> clone() const override {
+        return std::make_unique<CSVNodeCopier<T>>(
+            this->sharedState, this->pkIndex, this->copyDesc, this->columns, this->pkColumnID);
```

Comment:
remove `this->` and for other copiers.

Reply:
`this` is to access base class fields here.